### DISCUSSION TITLE
Applying non-standard AVRDUDE programmer arguments

### DISCUSSION THREAD

**Discussion by eyalabraham** [Alex079/vscode-avr-helper#34] [Q&A]
I need to use AVRDUDE "-i <delay>" argument for serial port bit-banging. I see that the extension accepts "Arguments" but it does not look like it takes anything that I add in that field.

When running directly from command line, this AVRDUDE options adds line like: "Setting isp clock delay        :  20" when using "-i 20"

Looking at the code the function getDeviceInfo() does not seem to pick up that field.

Any help and suggestions on this front, workaround or perhaps fix is required?

EA

**Comment by Alex079** [comment]
Hello,

Have you tried setting the arguments without blanks?
![image](https://user-images.githubusercontent.com/21042652/206046853-7a11ec31-3b7d-46d8-a69d-7c364ab46dbc.png)
![image](https://user-images.githubusercontent.com/21042652/206046879-9c7a1858-3223-4310-81ce-1af930721b28.png)
or
![image](https://user-images.githubusercontent.com/21042652/206047124-c287e5e5-1b8a-4d23-a143-843e22d34b0d.png)
![image](https://user-images.githubusercontent.com/21042652/206047157-cf68a205-4157-4a2e-8eb3-7a43e39d9980.png)

Regards

**Reply by eyalabraham** [reply]
Hi, yes, I tried both cases you describe.
The Error window the pops up from AVRDUDE shows:

```
Error: 
avrdude: Version 6.3-20171130
         Copyright (c) 2000-2005 Brian Dean, http://www.bdmicro.com/
         Copyright (c) 2007-2014 Joerg Wunsch

         System wide configuration file is "/etc/avrdude.conf"
         User configuration file is "/home/eyal/.avrduderc"
         User configuration file does not exist or is not a regular file, skipping

         Using Port                    : /dev/ttyS5
         Using Programmer              : dasa
         AVR Part                      : ATmega328P
         Chip Erase delay              : 9000 us
         PAGEL                         : PD7
         BS2                           : PC2
         RESET disposition             : dedicated
         RETRY pulse                   : SCK
         serial program mode           : yes
         parallel program mode         : yes
         Timeout                       : 200
         StabDelay                     : 100
         CmdexeDelay                   : 25
         SyncLoops                     : 32
         ByteDelay                     : 0
         PollIndex                     : 3
         PollValue                     : 0x53
         Memory Detail                 :

                                  Block Poll               Page                       Polled
           Memory Type Mode Delay Size  Indx Paged  Size   Size #Pages MinW  MaxW   ReadBack
           ----------- ---- ----- ----- ---- ------ ------ ---- ------ ----- ----- ---------
           eeprom        65    20     4    0 no       1024    4      0  3600  3600 0xff 0xff
           flash         65     6   128    0 yes     32768  128    256  4500  4500 0xff 0xff
           lfuse          0     0     0    0 no          1    0      0  4500  4500 0x00 0x00
           hfuse          0     0     0    0 no          1    0      0  4500  4500 0x00 0x00
           efuse          0     0     0    0 no          1    0      0  4500  4500 0x00 0x00
           lock           0     0     0    0 no          1    0      0  4500  4500 0x00 0x00
           calibration    0     0     0    0 no          1    0      0     0     0 0x00 0x00
           signature      0     0     0    0 no          3    0      0     0     0 0x00 0x00

         Programmer Type : SERBB
         Description     : serial port banging, reset=rts sck=dtr mosi=txd miso=cts

avrdude: AVR device not responding
avrdude: initialization failed, rc=-1
         Double check connections and try again, or use -F to override
         this check.


avrdude done.  Thank you.

```
When I use AVRDUDE command line with ```avrdude -v -pm328p -cdasa -P/dev/ttyS5 -i 20 -Uflash:r:avr.hex``` I get expected working results:

```
avrdude: Version 6.3-20171130
         Copyright (c) 2000-2005 Brian Dean, http://www.bdmicro.com/
         Copyright (c) 2007-2014 Joerg Wunsch

         System wide configuration file is "/etc/avrdude.conf"
         User configuration file is "/home/eyal/.avrduderc"
         User configuration file does not exist or is not a regular file, skipping

         Using Port                    : /dev/ttyS5
         Using Programmer              : dasa
         Setting isp clock delay        :  20
         AVR Part                      : ATmega328P
         Chip Erase delay              : 9000 us
         PAGEL                         : PD7
         BS2                           : PC2
         RESET disposition             : dedicated
         RETRY pulse                   : SCK
         serial program mode           : yes
         parallel program mode         : yes
         Timeout                       : 200
         StabDelay                     : 100
         CmdexeDelay                   : 25
         SyncLoops                     : 32
         ByteDelay                     : 0
         PollIndex                     : 3
         PollValue                     : 0x53
         Memory Detail                 :

                                  Block Poll               Page                       Polled
           Memory Type Mode Delay Size  Indx Paged  Size   Size #Pages MinW  MaxW   ReadBack
           ----------- ---- ----- ----- ---- ------ ------ ---- ------ ----- ----- ---------
           eeprom        65    20     4    0 no       1024    4      0  3600  3600 0xff 0xff
           flash         65     6   128    0 yes     32768  128    256  4500  4500 0xff 0xff
           lfuse          0     0     0    0 no          1    0      0  4500  4500 0x00 0x00
           hfuse          0     0     0    0 no          1    0      0  4500  4500 0x00 0x00
           efuse          0     0     0    0 no          1    0      0  4500  4500 0x00 0x00
           lock           0     0     0    0 no          1    0      0  4500  4500 0x00 0x00
           calibration    0     0     0    0 no          1    0      0     0     0 0x00 0x00
           signature      0     0     0    0 no          3    0      0     0     0 0x00 0x00

         Programmer Type : SERBB
         Description     : serial port banging, reset=rts sck=dtr mosi=txd miso=cts

avrdude: AVR device initialized and ready to accept instructions

Reading | ################################################## | 100% 0.01s

avrdude: Device signature = 0x1e950f (probably m328p)
avrdude: safemode: lfuse reads as 62
avrdude: safemode: hfuse reads as D9
avrdude: safemode: efuse reads as FF
avrdude: reading flash memory:

Reading | ################################################## | 100% 83.99s

avrdude: Flash is empty, resulting file has no contents.
avrdude: writing output file "avr.hex"

avrdude: safemode: lfuse reads as 62
avrdude: safemode: hfuse reads as D9
avrdude: safemode: efuse reads as FF
avrdude: safemode: Fuses OK (E:FF, H:D9, L:62)

avrdude done.  Thank you.

```

The extra switch produces ```Setting isp clock delay        :  20``` which inserts a 20uSec delay between pulses when bit-banging the programmer lines. When I poked around the code, I didn't see PROGRAMMER_ARGS used anywhere other that when there is an exception. The args don't seem to be included in the string that is formed as the command line by performFlashTask()

**Reply by Alex079** [reply]
Indeed, this looks like a bug. Thank you for reporting it.

**Reply by eyalabraham** [reply]
Great, glad I could contribute something even as small as this.
I will be happy to test a fix if you want. I decided to move to VS from Eclipse and your extension will be quite useful to me.
Also, if you are considering a fix, perhaps you can output the AVRDUDE command line string to the console as debug aid. The Eclipse plugin I'm using does that as well as dump the AVRDUDE output to the console.

**Reply by Alex079** [reply]
You can try the version 2.1.0.
I added logs as well, but VSCode only shows them in attached debugger or in developer tools (`Help`>`Toggle Developer Tools`>`Console`). If this turns out to be not enough, we can think about adding a dedicated output channel for AVR Helper in `View`>`Output` panel.

**Reply by eyalabraham** [reply]
The fix to AVRDUDE general parameters works perfectly! thank you for checking that in so quickly. I'm starting a small project and I'll use VS and this extension. The developer console also give good output and it is very useful.

PS, I noticed that when I populate AVR > Compiler: Arguments in the "User" space, and delete everything from the "Workspace" space then nothing gets added to the compiler command string. As if the empty "Workspace" list overrides the populated "User" list. What I expected (and probably don't understand the explanation https://code.visualstudio.com/docs/getstarted/settings ) is that "User" list is high level and will appear any time, whereas "Workspace" list get added to it to provide extra localized settings.

**Reply by Alex079** [reply]
I think I get what you mean. Could you please try instead of removing values just "reset" the particular setting at the particular level?
The trick is that default value is not taken when any modification on any level exists. Modifications are indicated by "_modified_" label next to the setting name and by a thin line to the left of the value block.

**Reply by Alex079** [reply]
and it looks like there is no settings combination from different levels, there is only overriding

**Reply by eyalabraham** [reply]
Yes are 100% correct, thank you for clarifying this point. I've never worked with VS at this level and I have a bit of a learning curve.
I think I am all set with a small workaround: it seems that the two JSON files ('c_cpp_properties' and 'settings') in the .vscode directory hold the setting the way I like to use the project environment. I found that copying them from a project that is set up the way I want to a new project has the effect of setting the destination project properly to my spec.
I started a new project and so far things work 100% the way I expect them to.
